### PR TITLE
[SPARK-3454] separate json endpoints for data in the UI

### PR DESCRIPTION
**Pull request by squito**
Exposes data available in the UI as json over http.  Key points:
- new endpoints, handled independently of existing XyzPage classes.  Root entrypoint is `JsonRootResource`
- Uses jersey + jackson for routing & converting POJOs into json
- tests against known results in `HistoryServerSuite`
- also fixes some minor issues w/ the UI -- synchronizing on access to `StorageListener` & `StorageStatusListener`, and fixing some inconsistencies w/ the way we handle retained jobs & stages.

### CODE REVIEW

**Comment by squito** [comment]
updated version of https://github.com/apache/spark/pull/4435

* test file names are sanitized to only use word characters
* removed giant test file, replaced with a much smaller one

**Comment by vanzin** [comment]
If the changes are only in the test files, looks good to me. nit: pr title has a typo ("separate").

File: .rat-excludes [hunk 1]
```diff
@@ -74,5 +74,12 @@ logs
 .*scalastyle-output.xml
 .*dependency-reduced-pom.xml
 known_translations
+json_expectation
+local-1422981759269/*
```
Comment (pwendell) [inline .rat-excludes]: Are these timestamps here? If so, maybe this shouldn't be committed to the repo?

Comment (pwendell) [inline .rat-excludes]: Ah related to test files, nevermind.

**Comment by pwendell** [comment]
Okay verified the filenames are fixed. Going to pull this in now since it's way past the feature deadline.

**Comment by squito** [comment]
thanks @pwendell !

**Pull request closed without merging**